rfcs/taskforce/0001-buildkiteci.md

```diff
@@ -0,0 +1,215 @@
+-- Optional decorations stack on top of the required bits
+-- This step will rerun twice before giving up
+-- It will upload artifacts under /tmp/artifacts/** to google cloud
+-- It depends on coda-build, genesis-ledger, and libp2p-helper steps
+-- Then the 4 required fields are given: a nice label for the UI, the
+--   command to run, a machine-readable key for further dependency graph
+--   manipulation, and a list of globs for granular test execution.
+-- Note: By making dirtyWhen a _required_ field, we set ourselves up for
+--       a world in which CI doesn't run on docs changes
+let packageDebStep =
+   decorate3
+	   (withFlaky 2)
+	   (withArtifacts "/tmp/artifacts/**")
+		 (withDeps [ "coda-build", "genesis-ledger", "libp2p-helper" ])
+	   Command::{
+	     label = ":box: Package deb",
+	     command = "make deb"
+	     key = "package-deb",
+	     dirtyWhen = [ "src/**" ]
+	   }
+```
+
+### Workflow of adding/changing/debugging pipelines
+
+Run buildkite agents locally or on manually spun up compute instances to test rules by targeting specific agents with tags and using the CLI [https://github.com/buildkite/cli](https://github.com/buildkite/cli)
+
+Liberal use of artifact uploading/downloading to keep build re-playable from different steps
+
+## This design solves the above issues
+
+P1 → self-hosted on GKE
+
+P2.1.1 → Parallel by default; opt-in to dependencies
+
+P2.1.2 → Rely on artifacts in gcloud which should be very quick to upload and download from within Gcloud infrastructure
+
+P2.2 → We can choose whatever machines we want
+
+P2.3 → We use dynamic pipelines to expand work based on files touched
+
+P3 → Manual adhoc template expansion is replaced with *automatic* expansion using Dhall
+
+P4 → BuildKite agents can be installed and tagged locally (or on your own instances) to iterate on builds on your own hardware before putting it into production
+
+P5 → Flakiness is explicit in the Dhall configuration
+
+## Mapping of resource utilization to tasks
+
+Big machines — 8core, 30GB of RAM
+
+- Builds
+- Heavy integration tests
+
+Light machines — 1core, 3.75GB of RAM
+
+- Linting
+- Small tools
+
+## Projected Spend
+
+We can run 20 Big machines for ~80% of what we're paying circle per month even if these 20 machines were on 100% of the time and not preemptible.
+
+This will also become much cheaper when we do autoscaling or if we use preemptible instances and support retries in case of preempt failure.
+
+If necessary we can look into getting this system to work on Azure to make better use of our credits, but Buildkite works much better with Gcloud and/or Aws. Azure credits will be easily used up by compute nodes during QA-net debugging.
+
+## Rough plan of moving over
+
+Work can be parallelized across working on the deploy infrastructure, pipeline creation, artifact management work
+
+1. Start with a super easy job (lint)
+2. Enable on both Circle and Buildkite and disable on circle only after 3 working days of PRs
```

Comment:
I like this plan. This should enable us to not break existing PRs when we change the required CI steps.